sympy/polys/rootconditions.py

```diff
@@ -0,0 +1,259 @@
+from __future__ import annotations
+
+from sympy.polys.densebasic import dup
+from sympy.polys.domains.domain import Er, Domain
+from sympy.external.gmpy import MPQ
+
+from sympy.polys.domains import EXRAW, QQ, PolynomialRing, RationalField
+from sympy.polys.rings import PolyElement
+from sympy.polys.densebasic import dup_convert
+from sympy.polys.densetools import dup_clear_denoms
+from sympy.core.mul import prod
+from sympy.core.exprtools import factor_terms
+from sympy.core.expr import Expr
+from sympy.simplify.simplify import signsimp
+from sympy.core.mul import Mul
+
+
+def dup_routh_hurwitz(f: dup[Er], K: Domain[Er]) -> list[Er]:
+    """
+    Computes the Routh Hurwitz criteria of ``f``.
+
+    The conditions, in general, are represented by a list of multivariate
+    polynomials which must be positive to ensure stability.
+
+    Note: This method assumes that the leading coefficient is non-zero.
+    In the opposite case, additional verification is required.
+
+    Depending on the domain, a different approach is used.
+    In non-numeric cases, the algorithm is modified to avoid divisions.
+
+    References
+    ==========
+
+    .. [1] G. Meinsma: Elementary proof of the Routh-Hurwitz test.
+           Systems & Control Letters, Volume 25, Issue 4, 1995, Pages 237-242,
+           https://courses.washington.edu/mengr471/resources/Routh_Hurwitz_Proof.pdf
+
+    """
+    if K.is_QQ:
+        return _dup_routh_hurwitz_qq(f, K) # type: ignore
+
+    elif K.is_ZZ or K.is_RR:
+        pq = dup_convert(f, K, QQ)
+        conds = _dup_routh_hurwitz_qq(pq, QQ)
+        return dup_convert(conds, QQ, K)
+
+    elif K.is_PolynomialRing:
+        return _dup_routh_hurwitz_poly(f, K) # type: ignore
+
+    elif K.is_FractionField:
+        _, pp = dup_clear_denoms(f, K, convert=True)
+        conds = _dup_routh_hurwitz_poly(pp, K) # type: ignore
+        return dup_convert(conds, K.get_ring(), K)
+
+    else:
+        pe = dup_convert(f, K, EXRAW)
+        conds = _dup_routh_hurwitz_exraw(pe) # type: ignore
+        return dup_convert(conds, EXRAW, K)
```

Comment:
I'm not sure about having a catch-all else here. There are domains that should probably not be allowed such as complex domains like `ZZ_I`, `QQ_I` and `CC`.